test_conformance/buffers/test_buffer_fill.cpp

```diff
@@ -726,44 +712,68 @@ int test_buffer_fill_struct( cl_device_id deviceID, cl_context context, cl_comma
 
     global_work_size[0] = (size_t)num_elements;
 
-    // Test with random offsets and fill sizes
-    for ( n = 0; n < 8; n++ ){
-        offset_elements = (size_t)get_random_float( 0.f, (float)(num_elements - 8), d );
-        fill_elements = (size_t)get_random_float( 8.f, (float)(num_elements - offset_elements), d );
-        log_info( "Testing random fill from offset %d for %d elements: \n", (int)offset_elements, (int)fill_elements );
-
-        pattern = (TestStruct *)malloc(ptrSize);
-        pattern->a = (cl_int)genrand_int32(d);
-        pattern->b = (cl_float)get_random_float( -FLT_MAX, FLT_MAX, d );
 
-        inptr = (TestStruct *)align_malloc(ptrSize * num_elements, min_alignment);
-        for ( j = 0; j < offset_elements; j++ ) {
-            inptr[j].a = 0;
-            inptr[j].b =0;
-        }
-        for ( j = offset_elements; j < offset_elements + fill_elements; j++ ) {
-            inptr[j].a = pattern->a;
-            inptr[j].b = pattern->b;
-        }
-        for ( j = offset_elements + fill_elements; j < (size_t)num_elements; j++ ) {
-            inptr[j].a = 0;
-            inptr[j].b = 0;
+    for (src_flag_id = 0; src_flag_id < NUM_FLAGS; src_flag_id++)
+    {
+        log_info("Testing with cl_mem_flags: %s\n",
+                 flag_set_names[src_flag_id]);
+
+        err = create_single_kernel_helper(context, &program, &kernel, 1,
+                                          &struct_kernel_code,
+                                          "read_fill_struct");
+        if (err)
+        {
+            log_error(" Error creating program for struct\n");
+            free((void *)pattern);
+            align_free((void *)inptr);
+            align_free((void *)hostptr);
+            free_mtdata(d);
+            return -1;
         }
 
-        hostptr = (TestStruct *)align_malloc(ptrSize * num_elements, min_alignment);
-        memset(hostptr, 0, ptrSize * num_elements);
+        // Test with random offsets and fill sizes
+        for (n = 0; n < 8; n++)
+        {
+            offset_elements =
+                (size_t)get_random_float(0.f, (float)(num_elements - 8), d);
+            fill_elements = (size_t)get_random_float(
+                8.f, (float)(num_elements - offset_elements), d);
+            log_info("Testing random fill from offset %d for %d elements: \n",
+                     (int)offset_elements, (int)fill_elements);
+
+            pattern = (TestStruct *)malloc(ptrSize);
```

Comment:
I suggest that we change this to be a non-pointer within the scope of this loop body, which would avoid all of the `free(pattern)` calls in all of the error cases. e.g.
```
TestStruct pattern;
pattern.a = ...;
...
clEnqueueFillBuffer(..., &pattern, ....)
```